google/cloud/firestore_v1/query.py

```diff
@@ -239,3 +242,96 @@ def on_snapshot(docs, changes, read_time):
         return Watch.for_query(
             self, callback, document.DocumentSnapshot, document.DocumentReference
         )
+
+
+class CollectionGroup(Query):
+    """Represents a Collection Group in the Firestore API.
+
+    This is a specialization of :class:`.Query` that includes all documents in the
+    database that are contained in a collection or subcollection of the given
+    parent.
+
+    Args:
+        parent (:class:`~google.cloud.firestore_v1.collection.CollectionReference`):
+            The collection that this query applies to.
+    """
+    def __init__(
+        self,
+        parent,
+        projection=None,
+        field_filters=(),
+        orders=(),
+        limit=None,
+        limit_to_last=False,
+        offset=None,
+        start_at=None,
+        end_at=None,
+        all_descendants=True,
+    ) -> None:
+        # TODO: Validate all_descendants?
+        super(Query, self).__init__(
+            parent=parent,
+            projection=projection,
+            field_filters=field_filters,
+            orders=orders,
+            limit=limit,
+            limit_to_last=limit_to_last,
+            offset=offset,
+            start_at=start_at,
+            end_at=end_at,
+            all_descendants=all_descendants,
+        )
+
+    def get_partitions(self, partition_count) -> Generator[Query, None, None]:
+        """TODO"""
+        # Validate query?
```

Comment:
I'm pretty sure the queries this will work on is pretty restricted. I'm not sure what all the restrictions are at this point. I can and probably will write some tests to probe what the limitations are, but they might also be written down somewhere.